### ISSUE TITLE
"Deadlock" on server side during cleaning of expired connections

### ISSUE THREAD

**Issue by mikayeln** [orientechnologies/orientdb#6768]
## OrientDB Version, operating system, or hardware.
- v2.2 SNAPSHOT[ ] - .8[x]

## Operating System
- [ x ] Linux

We have a large application with extensive read and write operations. Also for monitorings need we use the following JS function on server (GetConnections).

```
var server = com.orientechnologies.orient.server.OServerMain.server();
if(server != null) {    
    var cm = server.getClientConnectionManager();
    if(cm != null) {        
        cm.cleanExpiredConnections();
        var conns = cm.getConnections().toArray();
        var arr = [];
        arr.push("Connections:" + conns.length);        
        return arr;
    } else {
        print('cannot get cm from server');
    }
} else {
    print('cannot get server');
}
```

Application calls JS function above every n minutes. Orient Server also has mechanism to clean expired connections using Connection manager and cleanExpiredConnections() functions. A couple days ago we catched "deadlock" in orient. Stack trace:

```
 - com.orientechnologies.orient.core.index.sbtree.local.OSBTree.acquireAtomicExclusiveLock() @bci=5, line=933 (Interpreted frame)
 - com.orientechnologies.orient.core.index.engine.OSBTreeIndexEngine.acquireAtomicExclusiveLock(java.lang.Object) @bci=4, line=226 (Interpreted frame)
 - com.orientechnologies.orient.core.index.OIndexAbstract.acquireAtomicExclusiveLock(java.lang.Object) @bci=12, line=973 (Interpreted frame)
 - com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.lockIndexes(java.util.TreeMap) @bci=138, line=4295 (Compiled frame)
 - com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.commit(com.orientechnologies.orient.core.tx.OTransaction, java.lang.Runnable) @bci=465, line=1422 (Compiled frame)
 - com.orientechnologies.orient.core.tx.OTransactionOptimistic.doCommit() @bci=151, line=566 (Interpreted frame)
 - com.orientechnologies.orient.core.tx.OTransactionOptimistic.commit(boolean) @bci=51, line=106 (Interpreted frame)
 - com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.commit(boolean) @bci=179, line=2733 (Compiled frame)
 - com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.close() @bci=38, line=1209 (Compiled frame)
 - com.orientechnologies.orient.server.OClientConnection.close() @bci=33, line=70 (Compiled frame)
 - com.orientechnologies.orient.server.OClientConnectionManager.cleanExpiredConnections() @bci=303, line=111 (Compiled frame)
 - sun.reflect.NativeMethodAccessorImpl.invoke0(java.lang.reflect.Method, java.lang.Object, java.lang.Object[]) @bci=0 (Interpreted frame)
 - sun.reflect.NativeMethodAccessorImpl.invoke(java.lang.Object, java.lang.Object[]) @bci=87, line=57 (Interpreted frame)
 - sun.reflect.DelegatingMethodAccessorImpl.invoke(java.lang.Object, java.lang.Object[]) @bci=6, line=43 (Compiled frame)
 - java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) @bci=57, line=601 (Compiled frame)
 - sun.org.mozilla.javascript.internal.MemberBox.invoke(java.lang.Object, java.lang.Object[]) @bci=21, line=167 (Compiled frame)
 - sun.org.mozilla.javascript.internal.NativeJavaMethod.call(sun.org.mozilla.javascript.internal.Context, sun.org.mozilla.javascript.internal.Scriptable, sun.org.mozilla.javascript.internal.Scriptable, java.lang.Object[]) @bci=510, line=245 (Compiled frame)
 - sun.org.mozilla.javascript.internal.optimizer.OptRuntime.callProp0(java.lang.Object, java.lang.String, sun.org.mozilla.javascript.internal.Context, sun.org.mozilla.javascript.internal.Scriptable) @bci=24, line=119 (Compiled frame)
 - sun.org.mozilla.javascript.internal.gen._Unknown_source__9318._c_GetConnections_20(sun.org.mozilla.javascript.internal.gen._Unknown_source__9318, sun.org.mozilla.javascript.internal.Context, sun.org.mozilla.javascript.internal.Scriptable, sun.org.mozilla.javascript.internal.Scriptable, java.lang.Object[]) @bci=131, line=380 (Interpreted frame)
 - sun.org.mozilla.javascript.internal.gen._Unknown_source__9318.call(sun.org.mozilla.javascript.internal.Context, sun.org.mozilla.javascript.internal.Scriptable, sun.org.mozilla.javascript.internal.Scriptable, java.lang.Object[]) @bci=248 (Interpreted frame)
 - sun.org.mozilla.javascript.internal.ContextFactory.doTopCall(sun.org.mozilla.javascript.internal.Callable, sun.org.mozilla.javascript.internal.Context, sun.org.mozilla.javascript.internal.Scriptable, sun.org.mozilla.javascript.internal.Scriptable, java.lang.Object[]) @bci=7, line=429 (Compiled frame)
```

```
com.orientechnologies.orient.core.index.sbtree.local.OSBTree.acquireAtomicExclusiveLock() @bci=5, line=933 (Interpreted frame)
com.orientechnologies.orient.core.index.engine.OSBTreeIndexEngine.acquireAtomicExclusiveLock(java.lang.Object) @bci=4, line=226 (Interpreted frame)
com.orientechnologies.orient.core.index.OIndexAbstract.acquireAtomicExclusiveLock(java.lang.Object) @bci=12, line=973 (Interpreted frame)
com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.lockIndexes(java.util.TreeMap) @bci=138, line=4295 (Compiled frame)
com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.commit(com.orientechnologies.orient.core.tx.OTransaction, java.lang.Runnable) @bci=465, line=1422 (Compiled frame)
com.orientechnologies.orient.core.tx.OTransactionOptimistic.doCommit() @bci=151, line=566 (Interpreted frame)
com.orientechnologies.orient.core.tx.OTransactionOptimistic.commit(boolean) @bci=51, line=106 (Interpreted frame)
com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.commit(boolean) @bci=179, line=2733 (Compiled frame)
com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.close() @bci=38, line=1209 (Compiled frame)
com.orientechnologies.orient.server.OClientConnection.close() @bci=33, line=70 (Compiled frame)
com.orientechnologies.orient.server.OClientConnectionManager.cleanExpiredConnections() @bci=303, line=111 (Compiled frame)
com.orientechnologies.orient.server.OClientConnectionManager$1.run() @bci=4, line=64 (Compiled frame)
java.util.TimerThread.mainLoop() @bci=221, line=555 (Compiled frame)
java.util.TimerThread.run() @bci=1, line=505 (Interpreted frame)
```

Almost 20 threads were in this state.  At some point after all resources were locked, orient server stopped responding. 

I know that you released 2.2.10 and there should be no more problems with growing expired connections. But for investigation we need your expert opinion on this.

**Comment by mikayeln** [comment]
If you need more detailed stack traces I can provide them.

**Comment by laa** [comment]
Hi @mikayeln this issue is already fixed in 2.2.10 so merely upgrade to 2.2.11 (already) and issue will gone.
